src/main/java/sx/blah/discord/util/MessageHistoryBuilder.java

```diff
@@ -0,0 +1,412 @@
+			throw new IllegalArgumentException("End time is null!");
+		}
+		this.endTime = endTime;
+		this.includeEnd = inclusive;
+		return this;
+	}
+
+	/**
+	 * Specifies the maximum number of messages to get. Defaults to -1 (no limit).
+	 *
+	 * @param limit max number of messages to get
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder withMaxCount(int limit) {
+		if (limit < 0 && !lenient) throw new IllegalArgumentException("Limit is negative!");
+		this.limit = limit;
+		return this;
+	}
+
+	/**
+	 * Returns a stream of the message history.
+	 *
+	 * @return the stream of the message history
+	 * @see MessageHistoryIterator
+	 */
+	public Stream<IMessage> stream() {
+		buildRange();
+		return new MessageHistoryIterator(range).stream();
+	}
+
+	/**
+	 * Returns an iterator for this message history.
+	 * NOTE: This object can be put directly in a for-each loop without using this method.
+	 *
+	 * @return an Iterator instance for the message history
+	 */
+	@Override
+	public Iterator<IMessage> iterator() {
+		buildRange();
+		return new MessageHistoryIterator(range);
+	}
+
+	/**
+	 * Requests a {@link MessageHistory} object with the parameters set in this builder.
+	 * NOTE: This will not block the thread during execution. Use {@link MessageHistoryBuilder#get()} to block the
+	 * thread.
+	 *
+	 * @return the message history
+	 * @see MessageHistory
+	 */
+	public CompletableFuture<MessageHistory> request() {
+		return CompletableFuture.supplyAsync(this::get);
+	}
+
+	/**
+	 * Builds a {@link MessageHistory} object with the parameters set in this builder.
+	 * NOTE: This will block the thread until all messages are retrieved. Use {@link MessageHistoryBuilder#request()}
+	 * to request this message history without blocking.
+	 *
+	 * @return the message history
+	 * @see MessageHistory
+	 */
+	public MessageHistory get() {
+		if (limit == 0) {
+			messages = new ArrayList<>(); // Empty array list because limit = 0
+
+		} else {
+			// Get cache from the channel, sorted
+			cached = channel.messages.stream().sorted(MessageComparator.REVERSED).collect(Collectors.toList());
+
+			buildRange();
+
+			// Initialize messages
+			messages = limit < 0 ?
+					new ArrayList<>(MESSAGE_CHUNK_COUNT) :
+					new ArrayList<>(limit);
+
+			// Checks for special case for efficiency- if a range is reverse ordered (start earlier in history than end) but
+			// count unlimited, then it is equivalent to switching endpoints and reversing the collection afterwards
+			if (range.isChronological() && limit < 0) {
+				range = new MessageHistoryRange(channel,
+						new MessageHistoryRange.Endpoint(endTime, includeEnd),
+						start == null ?
+								new MessageHistoryRange.Endpoint(startTime, includeStart) :
+								new MessageHistoryRange.Endpoint(start, includeStart)
+				);
+
+				fetch();
+				Collections.reverse(messages);
+			} else {
+				fetch();
+			}
+		}
+
+		return new MessageHistory(messages);
+	}
+
+	private void buildRange() {
```

Comment:
Should this check if range has already been built? If so you might consider changing to `Lazy<MessageHistoryRange>`